### PR TITLE
docs: global fee adr

### LINKED ISSUE #2485: Docs: global fee ADR
## Summary
Gaia has a local module globalfee. In order to use  and maintain this module, an ADR doc needs to be provided to specify the design, requirements, and any other technical details.

## Proposal

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned
- [ ] Is a spike necessary to map out how the issue should be approached?

### PR DESCRIPTION
**Pull request by yaruwangway**
## Description

Closes: #2485

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct `docs:` prefix in the PR title
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] Confirmed the correct `docs:` prefix in the PR title
- [ ] Confirmed all author checklist items have been addressed 
- [ ] Confirmed that this PR only changes documentation
- [ ] Reviewed content for consistency
- [ ] Reviewed content for thoroughness
- [ ] Reviewed content for spelling and grammar
- [ ] Tested instructions (if applicable)

### CODE REVIEW

File: docs/adrs/adr-001-globalfee.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+- **Module Initialization Order**: The genutils module must be initialized before the globalfee module. This is due to the `DeliverGenTxs` in the genutils module, is called during `initGenesis`. This function executes `DeliverTx`, which subsequently calls the AnteHandle in FeeDecorator, triggering the fee check in `DeliverTx`.
+  To prevent the `DeliverGenTxs` go through a fee check, the initialization of the globalfee module should occur after the genutils module. This sequencing ensures that all necessary components are in place when the fee check occurs. See [Gaia Issue #2489](https://github.com/cosmos/gaia/issues/2489) for more context.
+  To prevent the `DeliverGenTxs` go through a fee check, the initialization of the globalfee module should occur after the genutils module. This sequencing ensures that all necessary components are in place when the fee check occurs. See [Gaia Issue #2489](https://github.com/cosmos/gaia/issues/2489) for more context.
```
Comment (sainoe) [inline docs/adrs/adr-001-globalfee.md]: These sentences are duplicates of the above.
```suggestion

```

File: docs/adrs/adr-001-globalfee.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+	// Get local minimum-gas-prices
+	localFees := GetMinGasPrice(ctx, int64(tx.GetGas()))
+
+	// Return combined fee requirements
+	return CombinedFeeRequirement(globalFees, localFees)
+}
+```
+
+- **Deterministic Bypass Parameters**: The decision whether a message can bypass minimum fee has to be deterministic as well. To ensure this, `BypassMinFeeMsgTypes` and `MaxTotalBypassMinFeeMsgGasUsage` parameters are moved to a persistent store.
```
Comment (sainoe) [inline docs/adrs/adr-001-globalfee.md]:
```suggestion
- **Deterministic Bypass Parameters**: The decision of whether a message can bypass the minimum fee has to be deterministic as well. To ensure this, `BypassMinFeeMsgTypes` and `MaxTotalBypassMinFeeMsgGasUsage` parameters are moved to a persistent store.
```

File: docs/adrs/adr-001-globalfee.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	end
+	subgraph feeCoin
+	feeCoins-->feeCoinsZeroDenom;
+	feeCoins-->feeCoinsNonZeroDenom;
+	end
+
+	feeCoinsZeroDenom-.feeCheck-Zero.->zeroCoinFeesDenomReq;
+	feeCoinsNonZeroDenom-.feeCheck-NonZero.->nonzeroCoinFeesDenomReq;
+```
+plese note, `feeCoins` does not contain zero coins, the feeCoins are split to according to the denoms in `zeroCoinFeesDenomReq` or in `nonZeroCoinFeesDenomReq`, if feeCoins contains coins not in both  `zeroCoinFeesDenomReq` and `nonZeroCoinFeesDenomReq`, the transaction should be rejected.
```
Comment (sainoe) [inline docs/adrs/adr-001-globalfee.md]:
```suggestion
Please note that `feeCoins` does not contain zero coins. The fee coins are split according to the denoms in `zeroCoinFeesDenomReq` or `nonZeroCoinFeesDenomReq`. If feeCoins contains coins not in both `zeroCoinFeesDenomReq` and `nonZeroCoinFeesDenomReq`, the transaction should be rejected.
```
This sentence should be exentended to clairify that if `len(feeCoins)==0` and  `len(zeroCoinFeesDenomReq)` the transaction pass.

File: docs/adrs/adr-001-globalfee.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+## Context
+
+The globalfee module was created to manage a parameter called `MinimumGasPricesParam`, which sets a network-wide minimum fee. The intention was to stop random denominations from entering fee collections and to reduce the time validators take to check a long list of transaction fees. However, the initial version of the globalfee module had some issues:
+
+- In the globalfee module, several SDK coins methods were redefined because of the allowance of zero-value coins in the `MinimumGasPricesParam`. The `MinimumGasPricesParam` is of `sdk.DecCoins` type. In the cosmos-sdk, `sdk.DecCoins` are [sanitized](https://github.com/cosmos/cosmos-sdk/blob/67f04e629623d4691c4b2e48806f7793a3aa211e/types/dec_coin.go#L160-L177) to remove zero-value coins. As a result, several methods from `sdk.Coins` were [redefined in the Gaia fee antehandler](https://github.com/cosmos/gaia/blob/890ab3aa2e5788537b0d2ebc9bafdc968340e0e5/x/globalfee/ante/fee_utils.go#L46-L104).
+
+- `BypassMinFeeMsgTypes` exists in `app.toml`, and each node can define it. It's not deterministic that whether a transaction containing bypass-messages will be exempt from fee charge from fee charge.
```
Comment (sainoe) [inline docs/adrs/adr-001-globalfee.md:15]:
```suggestion
- `BypassMinFeeMsgTypes` exists in `app.toml`, and each node can define it. It's not deterministic whether a transaction containing bypass-messages will be exempt from fee charge.
```

Comment (sainoe) [inline docs/adrs/adr-001-globalfee.md]:
```suggestion
- `BypassMinFeeMsgTypes` exists in `app.toml`, and each node can define it. It's not deterministic whether a transaction containing bypass-messages will be exempt from fee charge from fee charge.
```

**Comment by sainoe** [comment]
Good work @yaruwangway! Overall, I believe all the technical decisions are covered. Maybe we could rework a bit the diagram and the examples. See my comments.

File: docs/adrs/adr-001-globalfee.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	feeCoinsNonZeroDenom, feeCoinsZeroDenom := splitCoinsByDenoms(feeCoins, zeroCoinFeesDenomReq)
+
+```
+
+```go 
+	if !feeCoinsNonZeroDenom.IsAnyGTE(nonZeroCoinFeesReq) {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, "insufficient fees; got: %s required: %s", feeCoins.String(), feeRequired.String())
+	}
+```
+An example of coins split in fee antehandler:
```
Comment (sainoe) [inline docs/adrs/adr-001-globalfee.md]: I like the example but it's hard to read. Could please you split it it different lines?

File: docs/architecture/adr-002-globalfee.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
+# ADR 002: Globalfee Module
+
+## Changelog
+* 2023-06-12: Initial Draft
+
+## Status
+Globalfee module is introduced into gaia from v8.0.0, and is refactored in v11.0.0.
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md]:
```suggestion
ACCEPTED Implemented
```

File: docs/architecture/adr-002-globalfee.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+## Context
+
+The globalfee module was created to manage a parameter called `MinimumGasPricesParam`, which sets a network-wide minimum fee. The intention was to stop random denominations from entering fee collections and to reduce the time validators take to check a long list of transaction fees. However, the initial version of the globalfee module had some issues:
+
+- In the globalfee module, several SDK coins methods were redefined because of the allowance of zero-value coins in the `MinimumGasPricesParam`. The `MinimumGasPricesParam` is of `sdk.DecCoins` type. In the cosmos-sdk, `sdk.DecCoins` are [sanitized](https://github.com/cosmos/cosmos-sdk/blob/67f04e629623d4691c4b2e48806f7793a3aa211e/types/dec_coin.go#L160-L177) to remove zero-value coins. As a result, several methods from `sdk.Coins` were [redefined in the Gaia fee antehandler](https://github.com/cosmos/gaia/blob/890ab3aa2e5788537b0d2ebc9bafdc968340e0e5/x/globalfee/ante/fee_utils.go#L46-L104).
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md]: nit: use SDK and comos-sdk consistently (either one or the other). Probably better to use Cosmos SDK.

File: docs/architecture/adr-002-globalfee.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+- `BypassMinFeeMsgTypes` exists in `app.toml`, and each node can define it. It's not deterministic whether a transaction containing bypass-messages will be exempt from fee charge from fee charge.
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md]:
```suggestion
- `BypassMinFeeMsgTypes` exists in `app.toml`, which means each node can define its own value. Thus, it's not clear whether a transaction containing bypass-messages will be exempted from paying a fee.
```

File: docs/architecture/adr-002-globalfee.md [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+- The fee check logic from globalfee is only executed in `checkTx`. This could allow malicious validators to change the fee check code and let transactions that do not meet the fee requirement pass.
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md]:
```suggestion
- The fee check logic is only executed in `checkTx` (as `MinimumGasPricesParam` is a local param). This could enable malicious validators to change the fee check code and propose transactions that do not meet the fee requirement.
```

File: docs/architecture/adr-002-globalfee.md [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+## Decision
+To fix these problems, the globalfee module is updated in Gaia v11:
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md]:
```suggestion
To fix these problems, the following changes are added to the globalfee module:
```

File: docs/architecture/adr-002-globalfee.md [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+- The fee check uses the SDK coins' methods instead of the redefined methods.
+- `BypassMinFeeMsgTypes` and `MaxTotalBypassMinFeeMsgGasUsage` have been moved to be part of the globalfee module, this makes the bypass-msgs are recognized at the network level.
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md]:
```suggestion
- `BypassMinFeeMsgTypes` is a param of the globalfee module. As a result, the check whether a transaction can bypass the minimum fee is done on `deliverTx` (i.e., at the network level). In addition, to improve user experience, `MaxTotalBypassMinFeeMsgGasUsage` is introduced as a globalfee param. 
```

File: docs/architecture/adr-002-globalfee.md [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+- The fee check is now executed in both `deliverTx` and `checkTx`. This is to prevent malicious validators from changing the fee check logic and allowing any transactions to pass fee check.
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md]:
```suggestion
- The fee check is now executed in both `deliverTx` and `checkTx`. This is to prevent malicious validators from changing the fee check logic and allowing any transactions to pass fee check. As a consequence, `MinimumGasPrices` is introduced as a globalfee param. 
```

Comment (mpoke) [reply docs/architecture/adr-002-globalfee.md]: Explain the difference between `MinimumGasPrices` the globalfee param and the local param set in `app.toml`.

File: docs/architecture/adr-002-globalfee.md [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+
+## Consequences
+
+### ZeroCoins in `MinimumGasPricesParam`
+The allowance of zero coins in the `MinimumGasPricesParam` within the globalfee module implies that `CombinedFeeRequirement(globalFees, localFees)` also permits zero coins. Therefore, the `CombinedFeeRequirement` doesn't meet the requirements of certain sdk.Coins methods. For instance, the `DenomsSubsetOf` method requires coins that do not contain zero coins.
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md]: First motivate why zero coints are allowed.

File: docs/architecture/adr-002-globalfee.md [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+
+## Decision
+To fix these problems, the globalfee module is updated in Gaia v11:
+- The fee check uses the SDK coins' methods instead of the redefined methods.
+- `BypassMinFeeMsgTypes` and `MaxTotalBypassMinFeeMsgGasUsage` have been moved to be part of the globalfee module, this makes the bypass-msgs are recognized at the network level.
+- The fee check is now executed in both `deliverTx` and `checkTx`. This is to prevent malicious validators from changing the fee check logic and allowing any transactions to pass fee check.
+
+## Consequences
+
+### ZeroCoins in `MinimumGasPricesParam`
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md:28]: Move this section as a subsection of `## Decision`.

Comment (mpoke) [reply docs/architecture/adr-002-globalfee.md:28]: Also, what's the diff between `MinimumGasPricesParam` and `MinimumGasPrices`?

File: docs/architecture/adr-002-globalfee.md [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+The allowance of zero coins in the `MinimumGasPricesParam` within the globalfee module implies that `CombinedFeeRequirement(globalFees, localFees)` also permits zero coins. Therefore, the `CombinedFeeRequirement` doesn't meet the requirements of certain sdk.Coins methods. For instance, the `DenomsSubsetOf` method requires coins that do not contain zero coins.
+
+To address this, the `CombinedFeeRequirement` is split into zero and non-zero coins, forming `nonZeroCoinFeesReq` and `zeroCoinFeesDenomReq`. Similarly, the paid fees (feeCoins) are split into `feeCoinsNonZeroDenom` and `feeCoinsZeroDenom`, based on the denominations of `nonZeroCoinFeesReq` and `zeroCoinFeesDenomReq`.
+
+The split enable checking `feeCoinsNonZeroDenom` against `nonZeroCoinFeesReq`, and `feeCoinsZeroDenom` against
+`zeroCoinFeesDenomReq`. In the check of `feeCoinsNonZeroDenom` against `nonZeroCoinFeesReq`, the sdk coins' methods can be used since zero coins are removed from the `nonZeroCoinFeesReq`, while in the check `feeCoinsZeroDenom` against `zeroCoinFeesDenomReq`, only denoms need to be checked.
+
+```go
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md:56]: What's the purpose of this code box?

File: docs/architecture/adr-002-globalfee.md [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+	nonZeroCoinFeesReq, zeroCoinFeesDenomReq := getNonZeroFees(feeRequired)
+
+	// feeCoinsNonZeroDenom contains non-zero denominations from the feeRequired
+	// feeCoinsNonZeroDenom is used to check if the fees meets the requirement imposed by nonZeroCoinFeesReq
+	// when feeCoins does not contain zero coins' denoms in feeRequired
+	feeCoinsNonZeroDenom, feeCoinsZeroDenom := splitCoinsByDenoms(feeCoins, zeroCoinFeesDenomReq)
+
+```
+
+```go
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md:95]: ditto: What's the purpose of this code box?

File: docs/architecture/adr-002-globalfee.md (hunk 14 above)
Comment (mpoke) [reply docs/architecture/adr-002-globalfee.md:28]: This section needs to be better structured as it's very hard to follow. Currently it contains the following:
- three paragraphs describing the concept of zero coins 
- two code blocks
- an example of coins split
- a diagram 
- a paragraph explaining the diagram
- a workflow diagram

File: docs/architecture/adr-002-globalfee.md [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+### Positive
+This refactor results in code that is easier to maintain. It prevents malicious validators from escaping fee checks and make the bypass messages work at network level.
+### Negative
+The introduction of FeeDecorator has replaced the usage of `MempoolFeeDecorator` in the cosmos-sdk. Currently, if both FeeDecorator and MempoolFeeDecorator are added to the AnteDecorator chain, it will result in redundant checks. However, there's potential for FeeDecorator and MempoolFeeDecorator to become incompatible in the future, depending on updates to the cosmos-sdk.
+
+### Neutral
+
+## References
+
+* [global fee docs](../modules/globalfee.md)
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md]:
```suggestion
* [Documentation of the globalfee module](../modules/globalfee.md)
```

**Review by mpoke — CHANGES_REQUESTED** [review]
Partial review. See my comments bellow.

File: docs/architecture/adr-002-globalfee.md (hunk 15 above)
Comment (yaruwangway) [reply docs/architecture/adr-002-globalfee.md:56]: i feel its more clear by showing the code than describing it.

File: docs/architecture/adr-002-globalfee.md (hunk 16 above)
Comment (yaruwangway) [reply docs/architecture/adr-002-globalfee.md:95]: same as above

File: docs/architecture/adr-002-globalfee.md (hunk 14 above)
Comment (yaruwangway) [reply docs/architecture/adr-002-globalfee.md:28]: `MinimumGasPrices` is `MinimumGasPricesParam`, i corrected it.

File: docs/architecture/adr-002-globalfee.md (hunk 13 above)
Comment (yaruwangway) [reply docs/architecture/adr-002-globalfee.md]: added

File: docs/architecture/adr-002-globalfee.md [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+- **ZeroCoins in `MinimumGasPricesParam`:**\
+Refactor the fee check logics, in order to use the Cosmos SDK coins' methods instead of the redefined methods.
+- **Bypass Message Types:**\
+`BypassMinFeeMsgTypes` is refactored to be a param of the globalfee module, in order to make the bypass messages deterministic.
+- **Check Fees in `DeliverTx`:**\
+The fee check is factored to executed in both `DeliverTx` and `CheckTx`. This is to prevent malicious validators from changing the fee check logic and allowing any transactions to pass fee check. As a consequence, `MinimumGasPricesParam` is introduced as a globalfee param.
+
+### ZeroCoins in `MinimumGasPricesParam`
+#### Coins Split
+The allowance of zero coins in the `MinimumGasPricesParam` within the globalfee module implies that `CombinedFeeRequirement(globalFees, localFees)` also permits zero coins. Therefore, the `CombinedFeeRequirement` doesn't meet the requirements of certain `sdk.Coins` methods. For instance, the `DenomsSubsetOf` method requires coins that do not contain zero coins.
```
Comment (mpoke) [inline docs/architecture/adr-002-globalfee.md:31]: Add a short description of `CombinedFeeRequirement`.

Comment (mpoke) [reply docs/architecture/adr-002-globalfee.md:31]: This is actually described in a section bellow. Just move the description here.

**Review by mpoke — APPROVED** [review]
Nice work @yaruwangway. Just one comment bellow.

File: docs/architecture/adr-002-globalfee.md (hunk 18 above)
Comment (yaruwangway) [reply docs/architecture/adr-002-globalfee.md:31]: hi, updated.

File: docs/architecture/adr-002-globalfee.md [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+
+**split the paid fees**:\
+if `paidFee=[1uatom, 0.5photon]`,
+the `splitCoinsByDenoms` splits the paidFee into `feeCoinsZeroDenom=[1uatom]` (the same denom as zero coins in `combinedFeeRequirement`), and `feeCoinsNonZeroDenom=[0.5stake]`
+then `feeCoinsZeroDenom=[1uatom]` is checked by `nonZeroCoinFeesReq=[1photon, 1stake]`.
+
+Please note that `feeCoins` does not contain zero coins. The fee coins are split according to the denoms in `zeroCoinFeesDenomReq` or `nonZeroCoinFeesDenomReq`. If feeCoins contains coins not in both `zeroCoinFeesDenomReq` and `nonZeroCoinFeesDenomReq`, the transaction should be rejected. On the contrary, if  feeCoins' denoms are in either `zeroCoinFeesDenomReq` or `nonZeroCoinFeesDenomReq`, and `len(zeroCoinFeesDenomReq)!=0`, the transaction can directly pass, otherwise, the fee amount need to be checked.
+
+
+### Bymass Message Types
```
Comment (sainoe) [inline docs/architecture/adr-002-globalfee.md]:
```suggestion
### Bypass Message Types
```

File: docs/architecture/adr-002-globalfee.md [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+  To prevent the `DeliverGenTxs` go through a fee check, the initialization of the globalfee module should occur after the genutils module. This sequencing ensures that all necessary components are in place when the fee check occurs. See [Gaia Issue #2489](https://github.com/cosmos/gaia/issues/2489) for more context.
+
+## Consequences
+### Positive
+This refactor results in code that is easier to maintain. It prevents malicious validators from escaping fee checks and make the bypass messages work at network level.
+### Negative
+The introduction of FeeDecorator has replaced the usage of `MempoolFeeDecorator` in the Cosmos SDK. Currently, if both FeeDecorator and MempoolFeeDecorator are added to the AnteDecorator chain, it will result in redundant checks. However, there's potential for FeeDecorator and MempoolFeeDecorator to become incompatible in the future, depending on updates to the Cosmos SDK.
+
+### Neutral
```
Comment (sainoe) [inline docs/architecture/adr-002-globalfee.md]:
```suggestion
```

**Review by sainoe — APPROVED** [review]
LGTM!
